ocaml/typing/typemod.ml

```diff
@@ -3116,6 +3144,52 @@ let () =
 
 (* Typecheck an implementation file *)
 
+let type_implementation_params_and_body env ast loc = function
+    [] ->
+      let str, sg, names, shape, finalenv =
+        type_implementation_structure false env ast in
+      Timpl_structure str,
+      Unit_signature sg,
+      names,
+      shape,
+      finalenv
+  | param_names ->
+      let params, newenv, _ =
+        List.fold_right (fun param_name (params, env, subst) ->
+          match param_name with
+            | "()" ->
+                Types.Unit :: params, env, subst
+            | _ ->
+                (* CR lmaurer: This sets things up so that we resolve the
+                   parameter name to a persistent identifier but then use a
+                   substitution to rewrite all references to use a local
+                   identifier instead. This makes some sense if the idea is to
+                   use the mechanism for persistent modules, but in any case
+                   it's sufficiently weird that it should be explained here. *)
```

Comment:
If we can find a way to avoid doing substitutions of persistent identifiers then we should. If we think of these as open terms rather than functors, then the parameters which depend on other parameters should also be open terms. So if the second parameter depends on the first then the second parameter should also declare the first as it's parameter. That would mean that the second parameter would have essentially a functor/open type with a non-persistent identifier for the first parameter and that would be the one substituted rather than the persistent identifier.

Reply:
But the first parameter isn't a parameter to the second parameter, it's a name that occurs free, which I would think needs a different mechanism.

Reply:
Punted on this, as discussed